server/src/main/java/org/opensearch/cluster/routing/allocation/decider/SameShardAllocationDecider.java

```diff
@@ -94,6 +95,8 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
             return decision;
         }
         if (node.node() != null) {
+            DiscoveryNode sourceShardNode = allocation.nodes().get(shardRouting.currentNodeId());
```

Comment:
canAllocate method is triggered for unassigned shards and currentNodeId would be null for those cases. To shard relocating scenarios, you should override canMoveAway or canRebalance methods, thoughts?

Reply:
`canAllocate` will always be triggered before allocate shard to a target node, including new shard allocation and shard relocation scenarios. As for `canMoveAway` and `canRebalance`, they determine whether a shard can be move away from the current node. The SameShardAllocationDecider only takes effect when the target node is determined.